content/blog/2018/delayering-istio/delayering-istio.md

```diff
@@ -0,0 +1,216 @@
+
+I see the same pattern repeat itself with the connectivity of microservices, and networking in general.  Network has been going through the changes that physical servers have gone through a decade earlier.  New set of layers and constructs are being introduced.  They are being baked deep into the protocol stack and even silicon without adequately considering low-touch alternatives.  Perhaps there is a way to remove those additional layers altogether.
+
+{{< image width="25%" ratio="175%"
+    link="burger.jpg"
+    alt="What current network stack has become"
+    >}}
+
+I have been thinking about these problems for some time and believe that an approach similar in concept to containers can be applied to the network stack that would fundamentally simplify how application endpoints are connected across the complexity of many intermediate layers.  I have reapplied the same principles from the original work on containers to create [AppSwitch](http://appswitch.io).  Similar to the way containers provide an interface that applications can directly consume, AppSwitch plugs directly into well-defined and ubiquitous network API that applications currently use and directly connects application clients to appropriate servers, skipping all intermediate layers.  In the end, that's what networking is all about.
+
+Before going into the details of how AppSwitch promises to remove unnecessary layers from the Istio stack, let me give a very brief introduction to its architecture.  Further details are available at the [documentation](https://appswitch.readthedocs.io/en/latest/) page.
+
+## AppSwitch
+
+Not unlike the container runtime, AppSwitch consists of a client and a daemon that speak REST with each other.  Both client and the daemon are built as one self-contained binary, `ax`.  Client transparently plugs into the application and tracks its system calls related to network connectivity and notifies the daemon about their occurrences.  As an example, let’s say an application makes the `connect(2)` system call to the service IP of a Kubernetes service.  AppSwitch client intercepts the connect, nullifies it and notifies the daemon about its occurrence along with some context that includes the system call arguments.  Daemon would then handle the system call, perhaps by directly connecting to the Pod IP of the upstream server on behalf of the application.
```

Comment:
The details are a bit tedious but important in this particular case.  Otherwise, it's easy to assume that there is a connection between client and daemon.